File: code/client/web/src/Domain/messages/MessageValidationError.ts
export abstract class MessageValidationError {
    protected constructor(public readonly defaultMessage: string = 'Invalid message') {}

    static ContentBlank = new (class extends MessageValidationError {
        constructor() {
            super('Message content cannot be blank');
        }
    })();

    static ContentLength = class extends MessageValidationError {
        constructor(
            public readonly min: number,
            public readonly max: number,
        ) {
            super(`Message content must be between ${min} and ${max} characters`);
        }
    };

    toErrorMessage(): string {
        return this.defaultMessage;
    }
}

File: code/server/jvm/im-domain/src/main/kotlin/im/domain/messages/MessageValidationError.kt
package im.domain.messages

sealed class MessageValidationError(
    val defaultMessage: String = "Invalid message",
) {
    data object ContentBlank : MessageValidationError("Message content cannot be blank")

    data class ContentLength(
        val min: Int,
        val max: Int,
    ) : MessageValidationError("Message content must be between $min and $max characters")

    fun toErrorMessage(): String = defaultMessage
}

fun List<MessageValidationError>.toErrorMessage(): String = joinToString("\n") { it.toString() }
